### PR TITLE
ISPN-10957 Support index/query by key for remote queries

### PR DESCRIPTION
**Pull request by fax4ever**
https://issues.redhat.com/browse/ISPN-10957

### CODE REVIEW

File: query/src/test/java/org/infinispan/query/model/Game.java [hunk 1]
```diff
@@ -41,7 +41,7 @@ public String toString() {
             '}';
    }
 
-   @AutoProtoSchemaBuilder(includeClasses = {Game.class, NonIndexedGame.class})
+   @AutoProtoSchemaBuilder(includeClasses = {Game.class, NonIndexedGame.class, GameKey.class})
```
Comment (tristantarrant) [inline query/src/test/java/org/infinispan/query/model/Game.java:44]: Can be changed to use `@ProtoSchema`

File: api/src/main/java/org/infinispan/api/annotations/indexing/Indexed.java [hunk 2]
```diff
@@ -37,4 +37,28 @@
     */
    boolean enabled() default true;
 
+   /**
+    * Enables indexes / queries by keys.
+    *
+    * @return The fully qualified name of the type of entities used as keys
+    */
+   String keyEntity() default "";
```
Comment (tristantarrant) [inline api/src/main/java/org/infinispan/api/annotations/indexing/Indexed.java:45]: We could also have a `Class<?> keyClass` to make it a bit more "typesafe". These would be mutually exclusive

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/IndexedFieldProvider.java [hunk 3]
```diff
@@ -62,47 +62,49 @@ interface FieldIndexingMetadata {
...
-         return false;
-      }
+   }
 
-      @Override
-      public Object getNullMarker(String[] propertyPath) {
-         return null;
-      }
-   };
+   static <TM> FieldIndexingMetadata<TM> noIndexing() {
```
Comment (tristantarrant) [inline object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/IndexedFieldProvider.java:70]: Invoking this method will now instantiate a new `FieldIndexingMetadata` every time. Previously it was reusing the same static instance

**Review by tristantarrant — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/infinispan/api/annotations/indexing/Indexed.java (hunk 2 above)
Comment (fax4ever) [reply api/src/main/java/org/infinispan/api/annotations/indexing/Indexed.java:45]: The problem is that for the remote entity mapping we want to put here the Protobuf message type and we don't support the embedded case at the moment.

File: object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/IndexedFieldProvider.java (hunk 3 above)
Comment (fax4ever) [reply object-filter/src/main/java/org/infinispan/objectfilter/impl/syntax/IndexedFieldProvider.java:70]: Right! I'm going to make also the instance static.

File: query/src/test/java/org/infinispan/query/model/Game.java (hunk 1 above)
Comment (fax4ever) [reply query/src/test/java/org/infinispan/query/model/Game.java:44]: Right!

**Comment by fax4ever** [comment]
Thank you @tristantarrant for the review. I added this commit to make the no index field provider caches as static instance again: https://github.com/infinispan/infinispan/pull/12281/commits/86eaed5a3ea3781037da025ce966afcdd9670a5c.
If you have different implementation ideas, of course we can change it.

**Comment by tristantarrant** [comment]
@fax4ever please also include documentation about this feature

**Comment by fax4ever** [comment]
> @fax4ever please also include documentation about this feature

I tried to write it

File: documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc [hunk 4]
```diff
@@ -0,0 +1,49 @@
+[id='queries_by_keys_{context}']
+= Queries by keys
+
+If a key of a cache entry has an indexed type, it is possible to index the key fields as well the the value fields,
+so that also the formers could be used in the `Ickle` queries.
```
Comment (tristantarrant) [inline documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc:5]: Former cannot be pluralized. Reword as:
Use indexed types as keys to query caches by key using Ickle.

File: documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+This can be done specifying the fully qualified name of the ProtocolBuffer message type to use as key type
+in the `keyEntity` attribute of the `@Indexed` annotation.
+
+[NOTE]
+====
+This feature is at the moment available only with indexed remote queries.
```
Comment (tristantarrant) [inline documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc:12]: This feature is currently only available with indexed remote queries.

File: documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+.Use key properties in the Ickle queries
+[source,slq]
+----
+select s.key.column from model.Structure s where s.key.zone = 'z7'
+----
+
+If the value already has a property named `key`, the definition of the key entity could create a naming conflict
+with the properties.
+For this reason but also in general,
```
Comment (tristantarrant) [inline documentation/src/main/asciidoc/topics/proc_query_queries_by_keys.adoc:39]: Remove this part of the sentence and start it with "It is possible..."

**Comment by fax4ever** [comment]
Thanks @tristantarrant. I applied the changes to the doc.

**Comment by fax4ever** [comment]
Thank you @tristantarrant

**Pull request merged**